pdns/filterpo.cc

```diff
@@ -697,31 +697,58 @@ DNSName DNSFilterEngine::Zone::maskToRPZ(const Netmask& nm)
   }
   else {
     DNSName temp;
-    const auto str = addr.toString();
-    const auto len = str.size();
-    std::string::size_type begin = 0;
-
-    while (begin < len) {
-      std::string::size_type end = str.find(":", begin);
-      std::string sub;
-      if (end != string::npos) {
-        sub = str.substr(begin, end - begin);
-      }
-      else {
-        sub = str.substr(begin);
+    static_assert(sizeof(addr.sin6.sin6_addr.s6_addr) == sizeof(uint16_t) * 8);
+    uint16_t *src = (uint16_t*) &addr.sin6.sin6_addr.s6_addr;
+    std::array<uint16_t,8> elems;
+
+    // this routine was adopted from glibc's inet_ntop6, written by Paul Vixie
+    // because the RPZ spec (https://datatracker.ietf.org/doc/html/draft-vixie-dnsop-dns-rpz-00#section-4.1.1) says:
+    //
+    //    If there exists more than one sequence of zero-valued fields of
+    //    identical length, then only the last such sequence is compressed.
+    //    Note that [RFC5952] specifies compressing the first such sequence,
+    //    but our notation here reverses the order of fields, and so must also
+    //    reverse the selection of which zero sequence to compress.
+    //
+    // 'cur.len > best.len' from the original code is replaced by 'cur.len >= best.len', so the last-longest wins.
+
+    struct { int base, len; } best = {-1, 0}, cur = {-1, 0};
```

Comment:
some `size_t` (or `ssize_t`) also for the `i` and you can do away with the casts

Reply:
I tried that earlier and got
```
filterpo.cc: In static member function ‘static DNSName DNSFilterEngine::Zone::maskToRPZ(const Netmask&)’:
filterpo.cc:717:27: warning: comparison of integer expressions of different signedness: ‘ssize_t’ {aka ‘long int’} and ‘long unsigned int’ [-Wsign-compare]
     for (ssize_t i = 0; i < elems.size(); i++) {
                         ~~^~~~~~~~~~~~~~
```

Reply:
The few casts seemed like the best deal :)

Reply:
I was thinking `ssize_t` for the struct members and `size_t` for the loop vars, but oh well.